Build errors with sinclair_ac on ESPHome 2026.1.1

Hi,
when compiling the custom component sinclair_ac with a recent ESPHome version, several  appear and a series of breaking compilation errors occur.	
Can you help with this?

`INFO ESPHome 2026.1.1
INFO Reading configuration /config/esphome/gree.yaml...
INFO Updating https://github.com/piotrva/esphome_gree_ac.git@None
INFO Detected timezone 'Europe/Warsaw'
WARNING GPIO8 is a strapping PIN and should only be used for I/O with care.
Attaching external pullup/down resistors to strapping pins can cause unexpected failures.
See https://esphome.io/guides/faq/#why-am-i-getting-a-warning-about-strapping-pins
WARNING WiFi AP is configured but neither captive_portal nor web_server is enabled. The AP will not be usable for configuration or monitoring. Add 'captive_portal:' or 'web_server:' to your configuration.
INFO Generating C++ source...
INFO Compiling app... Build path: /data/build/gree
Processing gree (board: esp32-c3-devkitm-1; framework: espidf; platform: https://github.com/pioarduino/platform-espressif32/releases/download/55.03.35/platform-espressif32.zip)
--------------------------------------------------------------------------------
HARDWARE: ESP32C3 160MHz, 320KB RAM, 4MB Flash
 - contrib-piohome @ 3.4.4 
 - framework-espidf @ 3.50502.0 (5.5.2) 
 - tool-cmake @ 4.0.3 
 - tool-esp-rom-elfs @ 2024.10.11 
 - tool-esptoolpy @ 5.1.0 
 - tool-mklittlefs @ 3.2.0 
 - tool-ninja @ 1.13.1 
 - tool-scons @ 4.40801.0 (4.8.1) 
 - toolchain-riscv32-esp @ 14.2.0+20251107
Reading CMake configuration...
Dependency Graph
|-- noise-c @ 0.1.10
Compiling .pioenvs/gree/src/esphome/components/sinclair_ac/esppac.cpp.o
Compiling .pioenvs/gree/src/esphome/components/sinclair_ac/esppac_cnt.cpp.o
Compiling .pioenvs/gree/src/esphome/components/socket/socket.cpp.o
Compiling .pioenvs/gree/src/esphome/components/status_led/status_led.cpp.o
src/esphome/components/sinclair_ac/esppac_cnt.cpp: In member function 'void esphome::sinclair_ac::CNT::SinclairACCNT::send_packet()':
src/esphome/components/sinclair_ac/esppac_cnt.cpp:239:64: error: cannot convert 'esphome::StringRef' to 'const char*' in initialization
  239 |         const char* custom_fan_mode = this->get_custom_fan_mode();
      |                                       ~~~~~~~~~~~~~~~~~~~~~~~~~^~
      |                                                                |
      |                                                                esphome::StringRef
src/esphome/components/sinclair_ac/esppac.cpp: In member function 'void esphome::sinclair_ac::SinclairAC::update_swing_horizontal(const std::string&)':
src/esphome/components/sinclair_ac/esppac.cpp:138:41: warning: 'esphome::select::Select::state' is deprecated: Use current_option() instead of .state. Will be removed in 2026.7.0 [-Wdeprecated-declarations]
  138 |         this->horizontal_swing_select_->state != this->horizontal_swing_state_)
      |                                         ^~~~~
In file included from src/esphome/components/sinclair_ac/esppac.h:5,
                 from src/esphome/components/sinclair_ac/esppac.cpp:2:
src/esphome/components/select/select.h:39:15: note: declared here
   39 |   std::string state{};
      |               ^~~~~
src/esphome/components/sinclair_ac/esppac.cpp:138:41: warning: 'esphome::select::Select::state' is deprecated: Use current_option() instead of .state. Will be removed in 2026.7.0 [-Wdeprecated-declarations]
  138 |         this->horizontal_swing_select_->state != this->horizontal_swing_state_)
      |                                         ^~~~~
src/esphome/components/select/select.h:39:15: note: declared here
   39 |   std::string state{};
      |               ^~~~~
src/esphome/components/sinclair_ac/esppac.cpp:138:41: warning: 'esphome::select::Select::state' is deprecated: Use current_option() instead of .state. Will be removed in 2026.7.0 [-Wdeprecated-declarations]
  138 |         this->horizontal_swing_select_->state != this->horizontal_swing_state_)
      |                                         ^~~~~
src/esphome/components/select/select.h:39:15: note: declared here
   39 |   std::string state{};
      |               ^~~~~
src/esphome/components/sinclair_ac/esppac.cpp: In member function 'void esphome::sinclair_ac::SinclairAC::update_swing_vertical(const std::string&)':
src/esphome/components/sinclair_ac/esppac.cpp:149:39: warning: 'esphome::select::Select::state' is deprecated: Use current_option() instead of .state. Will be removed in 2026.7.0 [-Wdeprecated-declarations]
  149 |         this->vertical_swing_select_->state != this->vertical_swing_state_)
      |                                       ^~~~~
src/esphome/components/select/select.h:39:15: note: declared here
   39 |   std::string state{};
      |               ^~~~~
src/esphome/components/sinclair_ac/esppac.cpp:149:39: warning: 'esphome::select::Select::state' is deprecated: Use current_option() instead of .state. Will be removed in 2026.7.0 [-Wdeprecated-declarations]
  149 |         this->vertical_swing_select_->state != this->vertical_swing_state_)
      |                                       ^~~~~
src/esphome/components/select/select.h:39:15: note: declared here
   39 |   std::string state{};
      |               ^~~~~
src/esphome/components/sinclair_ac/esppac.cpp:149:39: warning: 'esphome::select::Select::state' is deprecated: Use current_option() instead of .state. Will be removed in 2026.7.0 [-Wdeprecated-declarations]
  149 |         this->vertical_swing_select_->state != this->vertical_swing_state_)
      |                                       ^~~~~
src/esphome/components/select/select.h:39:15: note: declared here
   39 |   std::string state{};
      |               ^~~~~
src/esphome/components/sinclair_ac/esppac.cpp: In member function 'void esphome::sinclair_ac::SinclairAC::update_display(const std::string&)':
src/esphome/components/sinclair_ac/esppac.cpp:160:32: warning: 'esphome::select::Select::state' is deprecated: Use current_option() instead of .state. Will be removed in 2026.7.0 [-Wdeprecated-declarations]
  160 |         this->display_select_->state != this->display_state_)
      |                                ^~~~~
src/esphome/components/select/select.h:39:15: note: declared here
   39 |   std::string state{};
      |               ^~~~~
src/esphome/components/sinclair_ac/esppac.cpp:160:32: warning: 'esphome::select::Select::state' is deprecated: Use current_option() instead of .state. Will be removed in 2026.7.0 [-Wdeprecated-declarations]
  160 |         this->display_select_->state != this->display_state_)
      |                                ^~~~~
src/esphome/components/select/select.h:39:15: note: declared here
   39 |   std::string state{};
      |               ^~~~~
src/esphome/components/sinclair_ac/esppac.cpp:160:32: warning: 'esphome::select::Select::state' is deprecated: Use current_option() instead of .state. Will be removed in 2026.7.0 [-Wdeprecated-declarations]
  160 |         this->display_select_->state != this->display_state_)
      |                                ^~~~~
src/esphome/components/select/select.h:39:15: note: declared here
   39 |   std::string state{};
      |               ^~~~~
src/esphome/components/sinclair_ac/esppac.cpp: In member function 'void esphome::sinclair_ac::SinclairAC::update_display_unit(const std::string&)':
src/esphome/components/sinclair_ac/esppac.cpp:171:37: warning: 'esphome::select::Select::state' is deprecated: Use current_option() instead of .state. Will be removed in 2026.7.0 [-Wdeprecated-declarations]
  171 |         this->display_unit_select_->state != this->display_unit_state_)
      |                                     ^~~~~
src/esphome/components/select/select.h:39:15: note: declared here
   39 |   std::string state{};
      |               ^~~~~
src/esphome/components/sinclair_ac/esppac.cpp:171:37: warning: 'esphome::select::Select::state' is deprecated: Use current_option() instead of .state. Will be removed in 2026.7.0 [-Wdeprecated-declarations]
  171 |         this->display_unit_select_->state != this->display_unit_state_)
      |                                     ^~~~~
src/esphome/components/select/select.h:39:15: note: declared here
   39 |   std::string state{};
      |               ^~~~~
src/esphome/components/sinclair_ac/esppac.cpp:171:37: warning: 'esphome::select::Select::state' is deprecated: Use current_option() instead of .state. Will be removed in 2026.7.0 [-Wdeprecated-declarations]
  171 |         this->display_unit_select_->state != this->display_unit_state_)
      |                                     ^~~~~
src/esphome/components/select/select.h:39:15: note: declared here
   39 |   std::string state{};
      |               ^~~~~
src/esphome/components/sinclair_ac/esppac_cnt.cpp: In member function 'bool esphome::sinclair_ac::CNT::SinclairACCNT::processUnitReport()':
src/esphome/components/sinclair_ac/esppac_cnt.cpp:619:45: error: cannot convert 'esphome::StringRef' to 'const char*'
  619 |         if (strcmp(this->get_custom_fan_mode(), newFanMode) != 0) hasChanged = true;
      |                    ~~~~~~~~~~~~~~~~~~~~~~~~~^~
      |                                             |
      |                                             esphome::StringRef
In file included from /data/cache/platformio/packages/toolchain-riscv32-esp/riscv32-esp-elf/include/c++/14.2.0/cstring:43,
                 from src/esphome/core/helpers.h:6,
                 from src/esphome/core/component.h:8,
                 from src/esphome/components/climate/climate.h:3,
                 from src/esphome/components/sinclair_ac/esppac_cnt.h:2,
                 from src/esphome/components/sinclair_ac/esppac_cnt.cpp:2:
/data/cache/platformio/packages/toolchain-riscv32-esp/riscv32-esp-elf/include/string.h:36:18: note:   initializing argument 1 of 'int strcmp(const char*, const char*)'
   36 | int      strcmp (const char *, const char *);
      |                  ^~~~~~~~~~~~
src/esphome/components/sinclair_ac/esppac.cpp: In member function 'void esphome::sinclair_ac::SinclairAC::set_vertical_swing_select(esphome::select::Select*)':
src/esphome/components/sinclair_ac/esppac.cpp:253:57: error: cannot convert 'esphome::sinclair_ac::SinclairAC::set_vertical_swing_select(esphome::select::Select*)::<lambda(const std::string&, size_t)>' to 'std::function<void(unsigned int)>&&'
  253 |     this->vertical_swing_select_->add_on_state_callback([this](const std::string &value, size_t index) {
      |                                                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                                         |
      |                                                         esphome::sinclair_ac::SinclairAC::set_vertical_swing_select(esphome::select::Select*)::<lambda(const std::string&, size_t)>
  254 |         if (value == this->vertical_swing_state_)
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~        
  255 |             return;
      |             ~~~~~~~                                      
  256 |         this->on_vertical_swing_change(value);
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~           
  257 |     });
      |     ~                                                    
src/esphome/components/select/select.h:81:60: note:   initializing argument 1 of 'void esphome::select::Select::add_on_state_callback(std::function<void(unsigned int)>&&)'
   81 |   void add_on_state_callback(std::function<void(size_t)> &&callback);
      |                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
src/esphome/components/sinclair_ac/esppac.cpp: In member function 'void esphome::sinclair_ac::SinclairAC::set_horizontal_swing_select(esphome::select::Select*)':
src/esphome/components/sinclair_ac/esppac.cpp:263:59: error: cannot convert 'esphome::sinclair_ac::SinclairAC::set_horizontal_swing_select(esphome::select::Select*)::<lambda(const std::string&, size_t)>' to 'std::function<void(unsigned int)>&&'
  263 |     this->horizontal_swing_select_->add_on_state_callback([this](const std::string &value, size_t index) {
      |                                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                                           |
      |                                                           esphome::sinclair_ac::SinclairAC::set_horizontal_swing_select(esphome::select::Select*)::<lambda(const std::string&, size_t)>
  264 |         if (value == this->horizontal_swing_state_)
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~        
  265 |             return;
      |             ~~~~~~~                                        
  266 |         this->on_horizontal_swing_change(value);
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~           
  267 |     });
      |     ~                                                      
src/esphome/components/select/select.h:81:60: note:   initializing argument 1 of 'void esphome::select::Select::add_on_state_callback(std::function<void(unsigned int)>&&)'
   81 |   void add_on_state_callback(std::function<void(size_t)> &&callback);
      |                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
src/esphome/components/sinclair_ac/esppac.cpp: In member function 'void esphome::sinclair_ac::SinclairAC::set_display_select(esphome::select::Select*)':
src/esphome/components/sinclair_ac/esppac.cpp:273:50: error: cannot convert 'esphome::sinclair_ac::SinclairAC::set_display_select(esphome::select::Select*)::<lambda(const std::string&, size_t)>' to 'std::function<void(unsigned int)>&&'
  273 |     this->display_select_->add_on_state_callback([this](const std::string &value, size_t index) {
      |                                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                                  |
      |                                                  esphome::sinclair_ac::SinclairAC::set_display_select(esphome::select::Select*)::<lambda(const std::string&, size_t)>
  274 |         if (value == this->display_state_)
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~        
  275 |             return;
      |             ~~~~~~~                               
  276 |         this->on_display_change(value);
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~           
  277 |     });
      |     ~                                             
src/esphome/components/select/select.h:81:60: note:   initializing argument 1 of 'void esphome::select::Select::add_on_state_callback(std::function<void(unsigned int)>&&)'
   81 |   void add_on_state_callback(std::function<void(size_t)> &&callback);
      |                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
src/esphome/components/sinclair_ac/esppac.cpp: In member function 'void esphome::sinclair_ac::SinclairAC::set_display_unit_select(esphome::select::Select*)':
src/esphome/components/sinclair_ac/esppac.cpp:283:55: error: cannot convert 'esphome::sinclair_ac::SinclairAC::set_display_unit_select(esphome::select::Select*)::<lambda(const std::string&, size_t)>' to 'std::function<void(unsigned int)>&&'
  283 |     this->display_unit_select_->add_on_state_callback([this](const std::string &value, size_t index) {
      |                                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                                       |
      |                                                       esphome::sinclair_ac::SinclairAC::set_display_unit_select(esphome::select::Select*)::<lambda(const std::string&, size_t)>
  284 |         if (value == this->display_unit_state_)
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~        
  285 |             return;
      |             ~~~~~~~                                    
  286 |         this->on_display_unit_change(value);
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~           
  287 |     });
      |     ~                                                  
src/esphome/components/select/select.h:81:60: note:   initializing argument 1 of 'void esphome::select::Select::add_on_state_callback(std::function<void(unsigned int)>&&)'
   81 |   void add_on_state_callback(std::function<void(size_t)> &&callback);
      |                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
*** [.pioenvs/gree/src/esphome/components/sinclair_ac/esppac_cnt.cpp.o] Error 1
*** [.pioenvs/gree/src/esphome/components/sinclair_ac/esppac.cpp.o] Error 1
========================== [FAILED] Took 7.63 seconds ==========================
`

--- a/components/sinclair_ac/esppac.cpp
+++ b/components/sinclair_ac/esppac.cpp
@@ -250,7 +250,11 @@ void SinclairAC::set_current_temperature_sensor(sensor::Sensor *current_temperat
 void SinclairAC::set_vertical_swing_select(select::Select *vertical_swing_select)
 {
     this->vertical_swing_select_ = vertical_swing_select;
-    this->vertical_swing_select_->add_on_state_callback([this](const std::string &value, size_t index) {
+    this->vertical_swing_select_->add_on_state_callback([this](size_t index) {
+        auto selected = this->vertical_swing_select_->at(index);
+        if (!selected.has_value())
+            return;
+        auto &value = selected.value();
         if (value == this->vertical_swing_state_)
             return;
         this->on_vertical_swing_change(value);
@@ -260,7 +264,11 @@ void SinclairAC::set_vertical_swing_select(select::Select *vertical_swing_select
 void SinclairAC::set_horizontal_swing_select(select::Select *horizontal_swing_select)
 {
     this->horizontal_swing_select_ = horizontal_swing_select;
-    this->horizontal_swing_select_->add_on_state_callback([this](const std::string &value, size_t index) {
+    this->horizontal_swing_select_->add_on_state_callback([this](size_t index) {
+        auto selected = this->horizontal_swing_select_->at(index);
+        if (!selected.has_value())
+            return;
+        auto &value = selected.value();
         if (value == this->horizontal_swing_state_)
             return;
         this->on_horizontal_swing_change(value);
@@ -270,7 +278,11 @@ void SinclairAC::set_horizontal_swing_select(select::Select *horizontal_swing_se
 void SinclairAC::set_display_select(select::Select *display_select)
 {
     this->display_select_ = display_select;
-    this->display_select_->add_on_state_callback([this](const std::string &value, size_t index) {
+    this->display_select_->add_on_state_callback([this](size_t index) {
+        auto selected = this->display_select_->at(index);
+        if (!selected.has_value())
+            return;
+        auto &value = selected.value();
         if (value == this->display_state_)
             return;
         this->on_display_change(value);
@@ -280,7 +292,11 @@ void SinclairAC::set_display_select(select::Select *display_select)
 void SinclairAC::set_display_unit_select(select::Select *display_unit_select)
 {
     this->display_unit_select_ = display_unit_select;
-    this->display_unit_select_->add_on_state_callback([this](const std::string &value, size_t index) {
+    this->display_unit_select_->add_on_state_callback([this](size_t index) {
+        auto selected = this->display_unit_select_->at(index);
+        if (!selected.has_value())
+            return;
+        auto &value = selected.value();
         if (value == this->display_unit_state_)
             return;
         this->on_display_unit_change(value);

--- a/components/sinclair_ac/esppac_cnt.cpp
+++ b/components/sinclair_ac/esppac_cnt.cpp
@@ -236,7 +236,7 @@ void SinclairACCNT::send_packet()
     bool    fanTurbo  = false;
     if (this->has_custom_fan_mode())
     {
-        const char* custom_fan_mode = this->get_custom_fan_mode();
+        const char* custom_fan_mode = this->get_custom_fan_mode().c_str();
 
         if (strcmp(custom_fan_mode, fan_modes::FAN_AUTO) == 0)
         {
@@ -616,7 +616,7 @@ bool SinclairACCNT::processUnitReport()
     const char* newFanMode = determine_fan_mode();
     if (this->has_custom_fan_mode())
     {
-        if (strcmp(this->get_custom_fan_mode(), newFanMode) != 0) hasChanged = true;
+        if (strcmp(this->get_custom_fan_mode().c_str(), newFanMode) != 0) hasChanged = true;
     }
     else
     {

